src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (size == 0) {
+            tail = newNode;
+            head = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (size == index) {
+            add(value);
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> newNode = new Node<>(current.prev, value, current);
+            newNode.next = current;
+            newNode.prev = current.prev;
+
+            if (current.prev != null) {
+                current.prev.next = newNode;
+            } else {
+                head = newNode;
+            }
+            current.prev = newNode;
+            size++;
+
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T oldVal = node.item;
+        node.item = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> removeNode = getNode(index);
+        if (removeNode.prev != null) {
+            removeNode.prev.next = removeNode.next;
+        } else {
+            head = removeNode.next;
+        }
+        if (removeNode.next != null) {
+            removeNode.next.prev = removeNode.prev;
+        } else {
+            tail = removeNode.prev;
+        }
+        size--;
+        return removeNode.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> nodeToRemove = findNode(object);
+        if (nodeToRemove == null) {
+            return false;
+        }
+        unlink(nodeToRemove);
+        return true;
+    }
+
+    private Node<T> findNode(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (current.item == object || (current.item != null && current.item.equals(object))) {
+                return current;
+            }
+            current = current.next;
+        }
+        return null;
+    }
+
+    private void unlink(Node<T> node) {
```

Comment:
Private methods and classes should be after public ones in your class. The `unlink` method should be moved below the public methods.